### PR TITLE
[rush] Allow the option to ignore dependencies when running bulk commands

### PR DESCRIPTION
**Pull request by aboktor**
… to allow running things like UT and Lint that don't require dependency order execution

### CODE REVIEW

**Comment by aboktor** [comment]
@christiango  FYI

File: apps/rush-lib/src/logic/TaskSelector.ts [hunk 1]
```diff
@@ -114,15 +115,17 @@ export class TaskSelector {
     }
   }
 
-  private _registerAll(): void {
+  private _registerAll(addDependencies: boolean): void {
+    console.log(addDependencies);
```
Comment (christiango) [inline apps/rush-lib/src/logic/TaskSelector.ts]: probably want to remove this

File: apps/rush-lib/src/logic/TaskSelector.ts [hunk 2]
```diff
@@ -114,15 +115,17 @@ export class TaskSelector {
     }
   }
 
-  private _registerAll(): void {
+  private _registerAll(addDependencies: boolean): void {
```
Comment (christiango) [inline apps/rush-lib/src/logic/TaskSelector.ts]: is this condition inverted here? The option being passed in is ignore dependencies, but the param here is add dependencies

Comment (aboktor) [reply apps/rush-lib/src/logic/TaskSelector.ts]: something went wrong here. Will upload a followup.

**Comment by aboktor** [comment]
@octogonz I have fixed up the PR, looks like I missed a few changes initially. Do you have any thoughts about how to improve this PR?

**Comment by octogonz** [comment]
I did have some thoughts about the design of this feature, but I'm on vacation right now and not spending any serious time looking at code. :-) I'll reply if I can find some time between mojitos; otherwise I get back on Thursday.

**Comment by christiango** [comment]
As a data point, lerna has this feature

File: common/changes/@microsoft/rush/fullParallel_2019-05-13-16-50.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "comment": "Adding ability to ignore dependencies for bulk commands",
```
Comment (octogonz) [inline common/changes/@microsoft/rush/fullParallel_2019-05-13-16-50.json]: Use an imperative verb "add". (See [these guidelines](https://rushjs.io/pages/best_practices/change_logs/).)

File: apps/rush-lib/src/schemas/command-line.schema.json [hunk 4]
```diff
@@ -72,6 +72,11 @@
               "title": "Ignore Missing Script",
               "description": "Normally Rush requires that each project's package.json has a \"scripts\" entry matching the custom command name. To disable this check, set \"ignoreMissingScript\" to true.",
               "type": "boolean"
+            },
+            "ignoreDependencies": {
+              "title": "Ignore dependency relation between projects and just run everything in parallel",
```
Comment (octogonz) [inline apps/rush-lib/src/schemas/command-line.schema.json]: The title should be "Ignore dependencies".  (Frankly I'm not sure we need these titles at all, but we should maintain consistency.)

**Comment by octogonz** [comment]
What is the behavior of this feature when used with the `--to` and `--from` parameters?  If I understand correctly, with `ignoreDependencies=true` then there is no operational difference between `--to` and `--from`, since the dependencies will not get processed at all.  Is that correct?  If so, is that an intuitive behavior?

Also, how would we describe the behavior of the `--to-version-policy` parameter?

(If people find that the `--to`, `--from`, `--to-version-policy`  parameters do not have intuitive interpretations in this mode, then perhaps a better design would be to introduce a new kind of command.  Currently we have "global commands" and "bulk commands".  Perhaps the dependency-ignoring mode could be called a "project command" or "unordered bulk command" or somesuch.)

**Comment by octogonz** [comment]
A commented example should be added to the `rush init` [command-line.json template](https://github.com/microsoft/web-build-tools/blob/master/apps/rush-lib/assets/rush-init/common/config/rush/command-line.json).  (But let's finalize the design first.)

**Comment by octogonz** [comment]
> (If people find that the `--to`, `--from`, `--to-version-policy` parameters do not have intuitive interpretations in this mode, then perhaps a better design would be to introduce a new kind of command. Currently we have "global commands" and "bulk commands". Perhaps the dependency-ignoring mode could be called a "project command" or "unordered bulk command" or somesuch.)

Another possible design would be to say that dependencies /are/ processed, but in an arbitrary order. In this case the option could be called `"unordered"=true` instead of `"ignoreDependencies"=true`. I suppose the design depends on your expected usage of this feature.

@iclanton

**Comment by christiango** [comment]
@iclanton can you provide guidance here on next steps? This is a feature we are missing dearly from lerna.

**Comment by aboktor** [comment]
> > (If people find that the `--to`, `--from`, `--to-version-policy` parameters do not have intuitive interpretations in this mode, then perhaps a better design would be to introduce a new kind of command. Currently we have "global commands" and "bulk commands". Perhaps the dependency-ignoring mode could be called a "project command" or "unordered bulk command" or somesuch.)
> 
> Another possible design would be to say that dependencies /are/ processed, but in an arbitrary order. In this case the option could be called `"unordered"=true` instead of `"ignoreDependencies"=true`. I suppose the design decision depends on your expected usage of this feature.
> 
> @iclanton

Our usage in this case is to run build in dependency order, then run unit tests, lint, automation etc. without respecting dependency order to increase parallelism. So we really don't care about behavior of --to and --from in that case. Despite that, we should come up with a clear definition of what they should do.
I do like the idea of a unordered bulk command. Potentially with a --targets option that would scope down to only certain projects (think you're writing code and want to run UT for the one project you're working on)

**Comment by iclanton** [comment]
> So we really don't care about behavior of --to and --from in that case. Despite that, we should come up with a clear definition of what they should do.

I think it would make the most sense to run the command for the subset of projects that would have been covered by the `--to` or `--from` parameter, but in an arbitrary order. That's what I'd expect it to do.

It seems like the feature you're looking for is more "ignore dependency order" than "ignore dependencies," so perhaps the configuration option and related variables should be `ignoreDependencyOrder`. That'd also require some implementation changes. Can you link to the docs for lerna's feature?

>Potentially with a --targets option that would scope down to only certain projects (think you're writing code and want to run UT for the one project you're working on)

Would you want to list out the projects you want the command to run if you specify `--targets`? Like `rush <command> --targets projectA,projectB,projectC`? Would it make more sense to specify that in the `command-line.json` configuration file? BTW - you can run a command on a single project by running `rushx <command>` in your project's directory, where `<command>` is the name of a script in your project's `package.json`.

**Comment by aboktor** [comment]
> I think it would make the most sense to run the command for the subset of projects that would have been covered by the `--to` or `--from` parameter, but in an arbitrary order. That's what I'd expect it to do.
> 
> It seems like the feature you're looking for is more "ignore dependency order" than "ignore dependencies," so perhaps the configuration option and related variables should be `ignoreDependencyOrder`. That'd also require some implementation changes. Can you link to the docs for lerna's feature?

I'm leaning more and more towards a new command type: 'unordered bulk command'. Mainly because --to and --from don't really make sense with this feature at all. We're asking to ignore dependency order, it doesn't make sense that we decide which projects to run based on the dependency graph. Or is there a usecase for that that I'm missing?

lerna supports a [--no-sort](https://github.com/lerna/lerna/tree/master/core/global-options#--no-sort) option on any command that ignores dependency order and runs with max parallelism.

**Comment by octogonz** [comment]
`ignoreDependencyOrder` is the same idea I was advocating above as `unordered`. I think your proposed name is even better. If we simply rename the option and update the docs, we can keep the PR pretty much in its current form. The concept is that the command needs to be run for all dependencies (and is not incremental), but the order in which it gets run does not matter, and relaxing the order speeds things up. That seems logically sound.

Later it may be useful to introduce a third command type that ingores dependencies entirely, as suggested above, but we should wait until we have a scenario that clearly motivates it. The `ignoreDependencyOrder` design seems to adequately address the needs being voiced right now.

**Comment by octogonz** [comment]
This discussion would benefit from some specific realistic examples of commands people would want to implement, and how the projects would be selected for those commands.

**Comment by octogonz** [comment]
Also I wonder if project selection could use the proposed tagging feature: https://github.com/microsoft/web-build-tools/issues/1241

**Comment by iclanton** [comment]
So the `--to` and `--from` flags wouldn't even exist on commands of this new type? I think that makes the most sense.

Do you want to update this PR to implement that?

**Comment by aboktor** [comment]
> So the `--to` and `--from` flags wouldn't even exist on commands of this new type? I think that makes the most sense.
> 
> Do you want to update this PR to implement that?

Alright, updated PR to do that. I explicitly had to tell --to and --from to do nothing.

File: apps/rush-lib/src/schemas/command-line.schema.json (hunk 4 above)
Comment (aboktor) [reply apps/rush-lib/src/schemas/command-line.schema.json]: This needs to be removed now.

**Comment by aboktor** [comment]
@iclanton @octogonz  Did you guys have any more thoughts on this PR?

**Comment by octogonz** [comment]
Sorry about the lack of follow-up. Ian & co were at a conference this week, and I have been really busy. I should have some time over the weekend. This is an important feature so we want to get it merged asap.

**Comment by octogonz** [comment]
@christiango @aboktor I finally had some time to think about the design of this feature in depth. (Last week was pretty crazy for me.)  It helps to keep in mind real world examples of how this feature might be used.  

Here's 3 that I came up with:

- `clean` - A command that "cleans" the projects (i.e. deletes all built output files)
- `lint` - A command that invokes tslint (and will work correctly even if the projects have not been built)
- `retest` - A command that runs unit tests (assuming the projects have already been built)

**Is it always safe to run these commands unordered?**  Yes for `clean`.  Probably yes for `lint` (unless TSLint semantic rules are used and depend on a type imported from another project).  No for `retest`.  So although the default behavior is unordered, there should probably be a command-line option to override that.

**Does it make sense to use `--to` and `--from`?**  Contrary to the earlier conversion,  I think it DOES make sense.  For example, I could say `rush clean --to my-project` if I want to ensure that everything upstream of my project will get rebuilt by `rush build`, without rebuilding it right now.  Or I could do `rush retest --from my-project` if I want to rerun all the unit tests affected by this project (but without rebuilding them).

> >Potentially with a --targets option that would scope down to only certain projects (think you're writing code and want to run UT for the one project you're working on)
>
> Would you want to list out the projects you want the command to run if you specify `--targets`? Like `rush <command> --targets projectA,projectB,projectC`? 

Explicitly specifying individual targets (`--targets projectA`) or tagged groups (`--targets :my_tag`  as proposed in https://github.com/microsoft/web-build-tools/issues/1241) definitely makes sense.  But it seems equally applicable to the normal "ordered" bulk commands.

Thus, I'm now proposing:

- We should treat `ignoreDependencyOrder` merely as a performance optimization for bulk commands (not as a substantially different kind of command)

- The config value `ignoreDependencyOrder` should be a default behavior, and we can provide a command-line parameter (`--ignore-dependency-order`) to override it

- If we want to further refine the selection of projects (e.g. `--targets`), that should be considered its own feature independent of `ignoreDependencyOrder`

If you guys like this design, I can probably find some time over the 3-day weekend to update the PR.

**Comment by iclanton** [comment]
Sorry - I was at SharePoint conference last week. I like @octogonz's proposal.

**Comment by octogonz** [comment]
Okay I'll see if I can update the PR branch with this idea.

**Comment by octogonz** [comment]
Okay I have pushed an update to the branch implementing this idea.  Essentially we're bringing back the initial approach, but making the semantics easier to understand, and adding logic to make it behave correctly with the `--to` and `--from` parameters.

> * If we want to further refine the selection of projects (e.g. `--targets`), that should be considered its own feature independent of `ignoreDependencyOrder`

@aboktor @christiango Are you able to use `ignoreDependencyOrder` without this proposed `--targets`  addition?  I think it's a valuable addition.  I'd prefer to handle it via a separate PR, but if it's important we can include it in this one.

**Comment by octogonz** [comment]
(BTW I appreciate you guys' patience with all the delays in getting this feature sorted out.  Thanks!)

**Comment by aboktor** [comment]
> Okay I have pushed an update to the branch implementing this idea. Essentially we're bringing back the initial approach, but making the semantics easier to understand, and adding logic to make it behave correctly with the `--to` and `--from` parameters.
> 
> > * If we want to further refine the selection of projects (e.g. `--targets`), that should be considered its own feature independent of `ignoreDependencyOrder`
> 
> @aboktor @christiango Are you able to use `ignoreDependencyOrder` without this proposed `--targets` addition? I think it's a valuable addition, however I'd prefer to handle it via a separate PR, but if it's important we can include it in this one.

At this point we can 'kinda' get the --target functionality by just running the npm command in the right folder. So what we are really interested in is the ignoreDependencyOrder. I'll take a look at the PR now.

**Comment by aboktor** [comment]
> LGTM, I don't have the permissions to 'Approve'

@octogonz  Any idea when we should expect this on npm?

**Comment by octogonz** [comment]
@aboktor this was published in 5.8.0 on June 11th

**Pull request merged**